File: data_structures/java/SLL.java
class node<obj>{//note this is a generic class
	node<obj> next;
	obj data;
	node(node<obj> n, obj a){
		next = n;
		data = a;
	}
	void setnext(node<obj> n){
		next = n;
	}
	void setele(obj a){
		data = a;
	}
	obj getele(){
		return data;
	}
	node<obj> getnext(){
		return next;
	}
}
class SLL<obj>{
	/*note this is a generic class to use it, the declaration will be
	SLL<Object(eg - Integer/String..)> Q = new SLL<Object>();
	*/
	private node<obj> head;
	private int size;
	SLL(){
		head = null;//sentinel node
		size=0;
	}
	void insertFirst(obj c){
		node<obj> n = new node<obj>(head, c);
		head=n;
		this.size++;
	}
	void removeFirst() throws NullPointerException{
		if(this.size == 0){
			throw new NullPointerException("SLL empty nothing to pop");
		}
		head = head.getnext();//the prev will be collected by garbage collector
		this.size--;		//since it has no pointer to it
		return;
	}
	node<obj> getNext(node n){
		return n.next;
	}
	int size(){
		return this.size;
	}
	Object get(int i) throws IndexOutOfBoundsException{//returns the obj at index i
		if(i>=size || i<0){
			throw new IndexOutOfBoundsException();
		}
		node<obj> temp = head;
		for (int j=0; j<i; j++){
			temp=temp.next;
		}
		return temp.data;
	}
	void insert(int posi, obj d) throws IndexOutOfBoundsException{//inserts the node at a particular index, throws error if index out of bound.
		if(posi>size){
			throw new IndexOutOfBoundsException("index for insert is greater then size");
		}else if(posi<0){
			throw new IndexOutOfBoundsException("index for insert is negative");
		}
		//0<= posi <= size. when equal to size insert a last.
		if(posi == 0 ){
			insertFirst(d);
			return;
		}
		node<obj> temp = head;
		node<obj> n = new node<obj>(head, d);
		while(posi>1){
			temp=temp.next;
			posi--;
		}
		n.next = temp.next;
		temp.next = n;
		this.size++;
		return;
	}
	void del(int posi) throws IndexOutOfBoundsException{
		if(posi>=size || posi<0){
			throw new IndexOutOfBoundsException();
		}
		if(posi==0){
			removeFirst();
			return;
		}
		node<obj> temp = head;
		node<obj> temp_prev = null;
		while(posi>0){
			temp_prev = temp;
			temp=temp.next;
			posi--;
		}
		temp_prev.next = temp.next;
		temp.next = null;
		size--;
		return;
	}
	int find(obj d){//returns the 0 based index in the of first ocuurence of obj d, else -1
		node<obj> temp = head;
		for(int i=0; i<size; i++){
			if(temp.data == d){
				return i;
			}
			temp = temp.next;
		}
		return -1;
	}
	boolean remove(obj d){//removes the first node with 'd' if exists
		node<obj> temp = head;
		node<obj> temp_prev = null;
		for(int i=0; i<size; i++){
			if(temp.data == d){
				break;
			}
			temp_prev = temp;
			temp = temp.next;
		}
		if(temp == head){
			removeFirst();
			return true;
		}else if(temp!=null){
			temp_prev.next = temp.next;
			size--;
			return true;
		}
		return false;
	}
}

File: data_structures/python/SLL.py
class node:
	def __init__(self, Node, data):
		self.next = Node
		self.data = data
	def setnext(self, n):
		self.next = n
	def getele(self):
		return self.data
	def setele(self, data):
		self.data = data
	def getnext(self):
		return self.next

class SLL:
	def __init__(self):
		self.head = None
		self.size = 0
	def insertFirst(self, D):
		n = node(self.head, D)
		self.head = n
		self.size +=1
	def removeFirst(self):
		if (self.size == 0):
			raise Exception("nothing to remove")
		else:
			self.head = self.head.getnext()
			self.size-=1
	def getNext(self, n):
		return n.next
	def get(self, i):
		if(i>=self.size or i<0):
			raise Exception("index out of bound")
		else:
			temp = self.head
			for i in range(i):
				temp=temp.next
			return temp.data
	def insert(self, posi, d):
		if(posi > self.size or posi <0):
			raise Exception("index out of bound")
		else:
			if(posi==0):
				self.insertFirst(d)
				return
			temp = self.head
			n = node(None, d)
			while(posi>1):
				temp = temp.next
				posi-=1
			n.next = temp.next
			temp.next = n
			self.size+=1
	def Del(self, posi):
		if(posi >= self.size or posi<0):
			raise Exception("index out of bound")
		else:
			if (posi==0):
				self.removeFirst()
				return
			temp = self.head
			temp_prev = None
			while (posi>0):
				temp_prev = temp
				temp = temp.next
				posi-=1
			temp_prev.next = temp.next
			temp.next = None
			self.size-=1
	def find(self, d):
		temp = self.head
		for i in range(self.size):
			if (temp.data == d):
				return i
			temp = temp.next
		return -1
	def remove(self, d):
		temp = self.head
		temp_prev = None
		for i in range(self.size):
			if(temp.data == d):
				break
			temp_prev = temp
			temp = temp.next
		if(temp == self.head):
			self.removeFirst()
			return True
		elif (temp!=None):
			temp_prev.next = temp.next
			self.size-=1
			return True
		else:
			return False

